drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,396 @@
+static bool init_irq = true;
+
+static void adc_gd32_isr(const struct device *dev)
+{
+	struct adc_gd32_data *data = dev->data;
+	const struct adc_gd32_config *cfg = dev->config;
+
+	if (ADC_STAT(cfg->reg) & ADC_STAT_EOC) {
+		*data->buffer++ = ADC_RDATA(cfg->reg);
+
+		/* Disable EOC interrupt. */
+		ADC_CTL0(cfg->reg) &= ~ADC_CTL0_EOCIE;
+		/* Clear EOC bit. */
+		ADC_STAT(cfg->reg) &= ~ADC_STAT_EOC;
+
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_gd32_data *data =
+		CONTAINER_OF(ctx, struct adc_gd32_data, ctx);
+	const struct device *dev = data->dev;
+	const struct adc_gd32_config *cfg = dev->config;
+
+	data->repeat_buffer = data->buffer;
+
+	/* Enable EOC interrupt */
+	ADC_CTL0(cfg->reg) |= ADC_CTL0_EOCIE;
+
+	/* Set ADC software conversion trigger. */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_SWRCST;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_gd32_data *data =
+		CONTAINER_OF(ctx, struct adc_gd32_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static inline void adc_gd32_calibration(const struct adc_gd32_config *cfg)
+{
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_RSTCLB;
+	/* Wait for calibration registers initialized. */
+	while (ADC_CTL1(cfg->reg) & ADC_CTL1_RSTCLB) {
+	}
+
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_CLB;
+	/* Wait for calibration complete. */
+	while (ADC_CTL1(cfg->reg) & ADC_CTL1_CLB) {
+	}
+}
+
+static int adc_gd32_configure_sampt(uint8_t channel, uint16_t acq_time)
+{
+	uint8_t index, offset;
+
+	/* Acquisitioin time unit is adc clock cycle. */
+	if (ADC_ACQ_TIME_UNIT(acq_time) != ADC_ACQ_TIME_TICKS) {
+		return -EINVAL;
+	}
+
+	index = 0;
+
+	if (acq_time != ADC_ACQ_TIME_DEFAULT) {
+		for (uint8_t i = 0; i < ARRAY_SIZE(acq_time_tbl); i++) {
+			if (ADC_ACQ_TIME_VALUE(acq_time) <= acq_time_tbl[i]) {
+				index = i;
+				break;
+			}
+		}
+
+		if (ADC_ACQ_TIME_VALUE(acq_time) != acq_time_tbl[index]) {
+			return -ENOTSUP;
+		}
+	}
+
+	if (channel < SAMPT1_SIZE) {
+		offset = SPT_WIDTH * channel;
+		ADC_SAMPT1(cfg->reg) &= ~(ADC_SAMPTX_SPTN << offset);
+		ADC_SAMPT1(cfg->reg) |= table_samp_time[index] << offset;
+	} else {
+		offset = SPT_WIDTH * (channel - SAMPT1_SIZE);
+		ADC_SAMPT0(cfg->reg) &= ~(ADC_SAMPTX_SPTN << offset);
+		ADC_SAMPT0(cfg->reg) |= table_samp_time[index] << offset;
+	}
+
+	return 0;
+}
+
+static int adc_gd32_channel_setup(const struct device *dev,
+			   const struct adc_channel_cfg *chan_cfg)
+{
+	const struct adc_gd32_config *cfg = dev->config;
+
+	if (chan_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Gain configuration not supported");
+		return -ENOTSUP;
+	}
+
+	if (chan_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Only ADC internal reference supported");
+		return -ENOTSUP;
+	}
+
+	if (chan_cfg->differential) {
+		LOG_ERR("differential sampling not supported");
```

Comment:
If you factor these (and other similar ones) to:

```c
LOG_ERR("%s not supported", "differential sampling");
```

you would be able to save quite a few bytes in ROM, since `%s not supported` is shared.

Reply:
I think it's OK, for a better read experience reason..

Reply:
I adjust the log message.
`gain` and `reference` will not give "not supported" message to user, instead with `invalid` message.